File: newspark-art/python/steps/login_steps.py
from behave import given, when, then
import time
from pages.login_page import LoginPage

@given('I am on the login page')
def step_impl(context):
    context.login_page.navigate()

@when('I login as "{user}"')
def step_impl(context, user):
    context.login_page.enter_credentials(user, user)

@when('I enter an incorrect username or password')
def step_impl(context):
    context.login_page.enter_credentials("admin", "wrong123")

@when('I click the Login button')
def step_impl(context):
    context.login_page.click_login_button()

@when('I click the Logout button')
def step_impl(context):
    context.login_page.click_login_button()

@then('I should be redirected to the "{page_name}" page')
def step_impl(context, page_name):
    context.page.wait_for_url(f"http://localhost:5173/{page_name}", timeout=3000)
    assert page_name in context.page.url, f"Expected page '{page_name}' in URL, but got '{context.page.url}'"

@then('I should see the login form')
def step_impl(context):
    assert context.login_page.username_input.is_visible(), "Login form not visible"

@then('I should see a Logout button')
def step_impl(context):
    assert context.login_page.is_logout_button_visible(), "Logout button not visible"

@given('I am logged in as "{user}"')
def step_impl(context, user):
    context.execute_steps(f'''
        Given I am on the login page
        When I login as "{user}"
        And I click the Login button
    ''')

@given('I am not logged in')
def step_impl(context):
    context.login_page.navigate()
    assert context.login_page.is_on_page(), "Not on login page"

@when('I navigate to the tasks page')
def step_impl(context):
    context.page.goto("http://localhost:5173/tasks")
    time.sleep(1)

File: newspark-art/TypeScript/tests/step-definitions/login.steps.ts
import { Given, When, Then } from '@cucumber/cucumber';
import { expect } from '@playwright/test';
import { CustomWorld } from '../support/world';

Given('I am on the login page', async function(this: CustomWorld) {
  await this.loginPage.navigate();
});

When('I login as {string}', async function(this: CustomWorld, user: string) {
  await this.loginPage.enterCredentials(user, user);
});

When('I enter an incorrect username or password', async function() {
  await this.loginPage.enterCredentials('admin', 'wrong123');
});

When('I click the Login button', async function() {
  await this.loginPage.clickLoginButton();
});

When('I click the Logout button', async function() {
  await this.loginPage.clickLogoutButton();
});

Then('I should be redirected to the {string} page', async function(pageName: string) {
  await this.page.waitForURL(`http://localhost:5173/${pageName}`, { timeout: 3000 });
  expect(this.page.url()).toContain(pageName);
});

Then('I should see the login form', async function() {
  const isVisible = await this.loginPage.isUsernameInputVisible();
  expect(isVisible).toBeTruthy();
});

Then('I should see a Logout button', async function() {
  const isVisible = await this.loginPage.isLogoutButtonVisible();
  expect(isVisible).toBeTruthy();
});

Given('I am logged in as {string}', async function(user: string) {
  // In TypeScript we execute steps sequentially rather than using execute_steps
  await this.loginPage.navigate();
  await this.loginPage.enterCredentials(user, user);
  await this.loginPage.clickLoginButton();
});

Given('I am not logged in', async function() {
  await this.loginPage.navigate();
  const isOnLoginPage = await this.loginPage.isOnPage();
  expect(isOnLoginPage).toBeTruthy();
});

When('I navigate to the tasks page', async function() {
  await this.page.goto('http://localhost:5173/tasks');
  // Converting Python's time.sleep to JavaScript's setTimeout wrapped in a Promise
  await new Promise(resolve => setTimeout(resolve, 1000));
});
